File: play-ws-standalone/src/main/java/play/libs/ws/DefaultBodyReadables.java
/*
 * Copyright (C) from 2022 The Play Framework Contributors <https://github.com/playframework>, 2011-2021 Lightbend Inc. <https://www.lightbend.com>
 */

package play.libs.ws;

import org.apache.pekko.stream.javadsl.Source;
import org.apache.pekko.util.ByteString;

import java.io.ByteArrayInputStream;
import java.io.InputStream;
import java.nio.ByteBuffer;

/**
 * This interface defines a set of reads for converting a response
 * into a readable format.
 */
public interface DefaultBodyReadables {

    /**
     * Converts a response body into an org.apache.pekko.util.ByteString:
     *
     * {{{
     * ByteString byteString = response.body(byteString())
     * }}}
     */
    default BodyReadable<ByteString> byteString() {
        return StandaloneWSResponse::getBodyAsBytes;
    }

    /**
     * Converts a response body into a String.
     *
     * Note: this is only a best-guess effort and does not handle all content types. See
     * {@link StandaloneWSResponse#getBody()} for more information.
     *
     * {{{
     * String string = response.body(string())
     * }}}
     */
    default BodyReadable<String> string() {
        return StandaloneWSResponse::getBody;
    }

    /**
     * Converts a response body into ByteBuffer.
     *
     * {{{
     * ByteBuffer buffer = response.body(byteBuffer())
     * }}}
     */
    default BodyReadable<ByteBuffer> byteBuffer() {
        // toByteBuffer returns a copy of the bytes
        return response -> response.getBodyAsBytes().toByteBuffer();
    }

    /**
     * Converts a response body into an array of bytes.
     *
     * {{{
     * byte[] byteArray = response.body(bytes())
     * }}}
     */
    default BodyReadable<byte[]> bytes() {
        return response -> response.getBodyAsBytes().toArray();
    }

    default BodyReadable<InputStream> inputStream() {
        return (response -> new ByteArrayInputStream(response.getBodyAsBytes().toArray()));
    }

    default BodyReadable<Source<ByteString, ?>> source() {
        return StandaloneWSResponse::getBodyAsSource;
    }

}

File: play-ws-standalone/src/main/scala/play/api/libs/ws/DefaultBodyReadables.scala
/*
 * Copyright (C) from 2022 The Play Framework Contributors <https://github.com/playframework>, 2011-2021 Lightbend Inc. <https://www.lightbend.com>
 */

package play.api.libs.ws

import java.nio.ByteBuffer

import org.apache.pekko.stream.scaladsl.Source
import org.apache.pekko.util.ByteString

/**
 * Defines common BodyReadable for a response backed by `org.asynchttpclient.Response`.
 */
trait DefaultBodyReadables {

  /**
   * Converts a response body into an `org.apache.pekko.util.ByteString`:
   *
   * {{{
   * import org.apache.pekko.util.ByteString
   * import play.api.libs.ws.DefaultBodyReadables._
   *
   * def example(response: play.api.libs.ws.StandaloneWSResponse): ByteString =
   *   response.body[ByteString]
   * }}}
   */
  implicit val readableAsByteString: BodyReadable[ByteString] = BodyReadable(_.bodyAsBytes)

  /**
   * Converts a response body into a `String`.
   *
   * Note: this is only a best-guess effort and does not handle all content types. See
   * [[StandaloneWSResponse.body:String*]] for more information.
   *
   * {{{
   * import play.api.libs.ws.DefaultBodyReadables._
   *
   * def example(response: play.api.libs.ws.StandaloneWSResponse): String =
   *   response.body[String]
   * }}}
   */
  implicit val readableAsString: BodyReadable[String] = BodyReadable(_.body)

  /**
   * Converts a response body into a read only `ByteBuffer`.
   *
   * {{{
   * import java.nio.ByteBuffer
   * import play.api.libs.ws.DefaultBodyReadables._
   *
   * def example(response: play.api.libs.ws.StandaloneWSResponse): ByteBuffer =
   *   response.body[ByteBuffer]
   * }}}
   */
  implicit val readableAsByteBuffer: BodyReadable[ByteBuffer] = BodyReadable(_.bodyAsBytes.asByteBuffer)

  /**
   * Converts a response body into `Array[Byte]`.
   *
   * {{{
   * import play.api.libs.ws.DefaultBodyReadables._
   *
   * def example(response: play.api.libs.ws.StandaloneWSResponse): Array[Byte] =
   *   response.body[Array[Byte]]
   * }}}
   */
  implicit val readableAsByteArray: BodyReadable[Array[Byte]] = BodyReadable(_.bodyAsBytes.toArray)

  /**
   * Converts a response body into `Source[ByteString, _]`.
   */
  implicit val readableAsSource: BodyReadable[Source[ByteString, ?]] = BodyReadable(_.bodyAsSource)
}

object DefaultBodyReadables extends DefaultBodyReadables
